Lib/test/test_import/__init__.py

```diff
@@ -790,6 +790,70 @@ def test_issue105979(self):
         self.assertIn("Frozen object named 'x' is invalid",
                       str(cm.exception))
 
+    def test_cwd_script_shadowing_stdlib(self):
+        with CleanImport('collections'):
+            import collections
+            collections.__spec__ = types.SimpleNamespace()
+            collections.__spec__.origin = os.path.join(os.getcwd(), 'collections.py')
+
+            with self.assertRaisesRegex(
+                AttributeError,
+                r"module 'collections' has no attribute 'does_not_exist' \(most "
+                r"likely due to '.*collections.py' shadowing the standard "
+                r"library module named 'collections'\)"
+            ):
+                collections.does_not_exist
+
+            del collections.__spec__.origin
+            with self.assertRaisesRegex(
+                AttributeError,
+                r"module 'collections' has no attribute 'does_not_exist'$"
+            ):
+                collections.does_not_exist
+            del collections.__spec__
+            with self.assertRaisesRegex(
+                AttributeError,
+                r"module 'collections' has no attribute 'does_not_exist'$"
+            ):
+                collections.does_not_exist
+
+    def test_shadowing_stdlib_sys_edge_cases(self):
+        program = textwrap.dedent('''
+import collections
+import os
+import types
+collections.__spec__ = types.SimpleNamespace()
+collections.__spec__.origin = os.path.join(os.getcwd(), 'collections.py')
+
+class substr(str):
+    __hash__ = None
+collections.__name__ = substr('collections')
+try:
+    collections.does_not_exist
+except TypeError as e:
+    print(str(e))
+
+import sys
+sys.stdlib_module_names = None
+try:
+    collections.does_not_exist
+except AttributeError as e:
+    print(str(e))
+
+del sys.stdlib_module_names
+try:
+    collections.does_not_exist
+except AttributeError as e:
+    print(str(e))
+''')
+        popen = script_helper.spawn_python('-c', program)
+        stdout, stderr = popen.communicate()
+        self.assertEqual(stdout.splitlines(), [
+            b"unhashable type: 'substr'",
```

Comment:
If I understand correctly, this had an AttributeError first, and then a TypeError was raised because the module's name isn't hashable. I feel like we should suppress the TypeError in this case and show the original exception. That's what happens today:

```
>>> import collections
>>> class substr(str): __hash__ = None
... 
>>> collections.__name__ = substr("hi")
>>> collections.hi
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    collections.hi
AttributeError: module 'hi' has no attribute 'hi'
```